app/controllers/password_resets_controller.rb

```diff
@@ -0,0 +1,79 @@
+class PasswordResetsController < ApplicationController
+
+  PASSWORD_RESET_PERMIT_PARAMS = %i[password password_confirmation].freeze
+
+  before_action :load_user_by_email, only: :create
+  before_action :load_user, :valid_user, :check_expiration,
+                only: %i(edit update)
+  before_action :check_empty_password, only: :update
+
+  # GET: /password_resets/new
+  def new; end
+
+  # PUT: /password_resets/:id/edit
+  def edit; end
+
+  def create
+    if @user
+      @user.create_reset_digest
+      @user.send_password_reset_email
+      flash[:info] = t("password_resets.email_sent")
+      redirect_to root_url
+    else
+      flash.now[:danger] = t("password_resets.email_not_found")
+      render :new, status: :unprocessable_entity
+    end
+  end
+
+  def update
+    if @user.update(user_params.merge(reset_digest: nil))
+      log_in @user
+      UserMailer.password_reset_confirmation(@user).deliver_now
+      flash[:success] = t("password_resets.password_reset_success")
+      redirect_to @user
+    else
+      render :edit, status: :unprocessable_entity
+    end
+  end
+
+  private
+
+  def user_params
+    params.require(:user).permit(*PASSWORD_RESET_PERMIT_PARAMS)
+  end
+
+  def load_user
+    @user = User.find_by(email: params[:email])
+    return if @user
+
+    flash[:danger] = t("password_resets.user_not_found")
+    redirect_to root_url
+  end
+
+  def valid_user
+    return if @user&.authenticated?(:reset, params[:id])
+
+    flash[:danger] = t("password_resets.invalid_user")
+    redirect_to root_url
+  end
+
+  def check_expiration
+    return unless @user.password_reset_expired?
+
+    flash[:danger] = t("password_resets.link_expired")
+    redirect_to new_password_reset_url
+  end
+
+  def load_user_by_email
+    email = params.dig(:password_reset, :email)&.downcase
+    @user = User.find_by(email: email)
```

Comment:
Cứ load thì phải xử lý cho trường hợp k tìm thấy chứ. Để các logic khác k cần check bằng toán tử an toàn nữa
Đưa nhánh else của hàm create vào check chỗ này

      flash.now[:danger] = t("password_resets.email_not_found")
      render :new, status: :unprocessable_entity